File: projectNew/old_src/Player.java
package com.f14.bg.player;

import com.f14.F14bg.consts.CmdConst;
import com.f14.bg.BGConst;
import com.f14.bg.action.BgResponse;
import com.f14.bg.chat.Message;
import com.f14.bg.common.ParamCache;
import com.f14.bg.component.Convertable;
import com.f14.bg.hall.User;
import com.f14.bg.report.Printable;

import java.util.HashMap;
import java.util.Map;

public abstract class Player implements Convertable, Printable {

    public User user;
    public int position = BGConst.INT_NULL;

    private ParamCache params = new ParamCache();
    private int team;

    public String getName() {
        return this.user.getName();
    }


    public ParamCache getParams() {
        return params;
    }

    public int getPosition() {
        return position;
    }

    @Override
    public String getReportString() {
        return String.format("玩家%d[%s]", this.position + 1, this.getName());
    }

    public int getTeam() {
        return team;
    }

    public void setTeam(int team) {
        this.team = team;
    }

    /**
     * 重置玩家的游戏信息
     */
    public void reset() {
        this.params.clear();
    }

    /**
     * 向玩家发送异常信息
     *
     * @param roomId
     * @param e
     */
    public void sendException(int roomId, Exception e) {
        user.getHandler().sendCommand(CmdConst.EXCEPTION_CMD, roomId, e.getMessage());
    }

    /**
     * 发送消息
     *
     * @param roomId
     * @param message
     */
    public void sendMessage(int roomId, Message message) {
        user.sendMessage(roomId, message);
    }

    /**
     * 发送回应
     *
     * @param roomId
     * @param res
     */
    public void sendResponse(int roomId, BgResponse res) {
        user.sendResponse(roomId, res);
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> res = new HashMap<>();
        res.put("userId", this.user.getId());
        res.put("chinese", this.getName());
        res.put("position", this.getPosition());
        res.put("team", this.getTeam());
        return res;
    }
}

File: projectNew/src/com/f14/bg/player/Player.kt
package com.f14.bg.player

import com.f14.F14bg.consts.CmdConst
import com.f14.bg.BGConst
import com.f14.bg.action.ISendableWith
import com.f14.bg.chat.IChatableWith
import com.f14.bg.common.ParamCache
import com.f14.bg.component.Convertable
import com.f14.bg.hall.GameRoom
import com.f14.bg.hall.User
import com.f14.bg.report.Printable

/**
 * Created by 吹风奈奈 on 2017/7/19.
 */
abstract class Player(var user: User, val room: GameRoom) : Convertable, Printable, ISendableWith<Int> by user, IChatableWith<Int> by user {
    var position = BGConst.INT_NULL
    var team: Int = 0
    val params = ParamCache()
    var name = user.name

//    val name: String
//        get() = this.user.name

    override val reportString
        get() = "玩家${this.position + 1}[${this.name}]"


    /**
     * 重置玩家的游戏信息
     */
    open fun reset() {
        this.params.clear()
    }

    /**
     * 向玩家发送异常信息
     * @param roomId
     * @param e
     */
    fun sendException(roomId: Int, e: Exception) {
        user.handler.sendCommand(CmdConst.EXCEPTION_CMD, roomId, e.message)
    }

//    /**
//     * 发送消息
//     * @param arg
//     * @param message
//     */
//    override fun sendMessage(arg: Int, message: Message) {
//        user.sendMessage(arg, message)
//    }
//
//    /**
//     * 发送回应
//     * @param arg
//     * @param res
//     */
//    override fun sendResponse(arg: Int, res: BgResponse) {
//        user.sendResponse(arg, res)
//    }

    override fun toMap(): Map<String, Any> = mapOf("userId" to this.user.id, "name" to this.name, "position" to this.position, "team" to this.team)

    fun isPrivatePosition(position: Int) = this.position == position || this.position < 0 && this.user.isAdmin
}
